WordPress/Tests/Commenting/HooksInlineDocsUnitTest.inc

```diff
@@ -0,0 +1,108 @@
+<?php
+// phpcs:set WordPress.Commenting.HooksInlineDocs allowed_extra_versions[] MU
+
+/*
+ * Set 1
+ *
+ * Make sure we check each function. These should all fail as there are no inline docs.
+ */
+
+apply_filters( 'hook', 'function' );
+apply_filters_ref_array( 'hook', 'function' );
+apply_filters_deprecated( 'hook', 'function' );
+do_action( 'hook', 'function' );
+do_action_ref_array( 'hook', 'function' );
+do_action_deprecated( 'hook', 'function' );
+
+/*
+ * Set 2
+ *
+ * Invalid comment styles. These should fail.
+ */
+
+// Wrong format.
+apply_filters( 'hook', 'function' );
+
+/*
+ * Docblocks need the /** format
+ */
+apply_filters( 'hook', 'function' );
+
+/*
+ * Set 3
+ *
+ * Docblocks missing required values. These should fail.
+ */
+
+// Has a since tag with no value.
+/**
+ * This is a description.
+ *
+ * @since
+ */
+apply_filters( 'hook', 'function' );
+
+// Has a since tag with an incomplete value.
+/**
+ * This is a description.
+ *
+ * @since 8.5
+ */
+apply_filters( 'hook', 'function' );
+
+// Has a since tag with an wrong value.
+/**
+ * This is a description.
+ *
+ * @since Test
+ */
+apply_filters( 'hook', 'function' );
+
+/*
+ * Set 4
+ *
+ * Valid comments. These should pass.
+ */
+
+// Complete docblock with a "normal" version number.
+/**
+ * This is a description.
+ *
+ * @since 1.0.0
+ */
+apply_filters( 'hook', 'function' );
+
+// Complete docblock with a "exceptional" version number declared in the XML.
+/**
+ * This is a description.
+ *
+ * @since MU
+ */
+apply_filters( 'hook', 'function' );
+
+// Complete docblock with a "normal" version number and extra text.
+/**
+ * This is a description.
+ *
+ * @since 1.0.0
+ * @since 1.2.0 Added something.
+ */
+apply_filters( 'hook', 'function' );
+
+// Complete docblock with a "exceptional" version number declared in the XML with extra text.
+/**
+ * This is a description.
+ *
+ * @since MU Additional text.
+ */
+apply_filters( 'hook', 'function' );
+
+// "Already documented" example of an action.
+/** This action is documented in file.php */
+do_action( 'hook ' );
+
+// "Already documented" example of a filter.
+/** This filter is documented in file.php */
+apply_filters(' hook', 'function' );
+
```

Comment:
These are good, but some real-world edge cases will be needed!

Consider:

```php
/**
 * This is my function DocBlock.
 *
 * @since 0000.00000.0000000000
 */
function my_function_has_a_hook() {
    do_action( 'foo' );
}
```

Likewise:

```php
/**
 * This is my function DocBlock.
 *
 * @since 0000.00000.0000000000
 */
function my_function_has_a_hook() {
    echo 'foo';
}
do_action( 'foo' );
```

If these both already get reported as missing the hook DocBlock, then great! Add them in as unit tests, and then try:

```php
/**
 * This is my function DocBlock.
 *
 * @since 0000.00000.0000000000
 */
function my_function_has_a_hook() { do_action( 'foo' ); }
```

😈

Reply:
The first two did catch "correctly", but probably wouldn't after I resolve the preceding line issue mentioned in https://github.com/WordPress/WordPress-Coding-Standards/pull/1892#discussion_r424599130

But, the third example of a single-line function fails the unit test since it doesn't see that as an error.

I'm going to push up what I have (and mark this as resolved since the tests are added) and think on it a bit.

I'm thinking for the third example, we can search for a comment previous to the hook but stop at a function declaration at least. May be others.

Thanks!

Reply:
ad220fdb

Reply:
Either here or PHPCSUtils, I bet @jrfnl has examples / helper functions to find something within a specific scope i.e. find a DocComment within the current scope.

Reply:
I do have docblock finder helper functions lined up to go into PHPCSUtils at some point, but as hook docblocks are a bit of a different beast compared to most structural elements I don't think that will help here.

I do have a draft `HooksTrait` lined up for WPCS 3.0.0 with logic for finding hook docblocks (for another sniff).

When I pull that, I will combine the logic for finding hook docblocks from this sniff with the logic in the trait to get the best of both brains in the trait (and switch this sniff over to using the trait).